psi4/src/psi4/libfock/DFJCOSK.cc

```diff
@@ -0,0 +1,1014 @@
+
+    // screening threshold
+    double thresh2 = options_.get_double("INTS_TOLERANCE") * options_.get_double("INTS_TOLERANCE");
+
+    // per-thread G Vector buffers (for accumulating thread contributions to G)
+    // G is the contraction of the density matrix with the 3-index ERIs
+    std::vector<std::vector<SharedVector>> GT(njk, std::vector<SharedVector>(nthreads_));
+
+    // H is the contraction of G with the inverse coulomb metric
+    std::vector<SharedVector> H(njk);
+
+    // per-thread J Matrix buffers (for accumulating thread contributions to J)
+    std::vector<std::vector<SharedMatrix>> JT(njk, std::vector<SharedMatrix>(nthreads_));
+
+    // initialize per-thread objects
+    for(size_t jki = 0; jki < njk; jki++) {
+        for(size_t thread = 0; thread < nthreads_; thread++) {
+            JT[jki][thread] = std::make_shared<Matrix>(nbf, nbf);
+            GT[jki][thread] = std::make_shared<Vector>(nbf_aux);
+        }
+        H[jki] = std::make_shared<Vector>(nbf_aux);
+    }
+
+    // diagonal shell maxima of J_metric_ for screening
+    std::vector<double> J_metric_shell_diag(nshell_aux, 0.0);
+    for (size_t s = 0; s < nshell_aux; s++) {
+        int bf_start = auxiliary_->shell(s).function_index();
+        int bf_end = bf_start + auxiliary_->shell(s).nfunction();
+        for (size_t bf = bf_start; bf < bf_end; bf++) {
+            J_metric_shell_diag[s] = std::max(J_metric_shell_diag[s], J_metric_->get(bf, bf));
+        }
+    }
+
+    // shell maxima of D for screening
+    Matrix Dshell(nshell, nshell);
+    auto Dshellp = Dshell.pointer();
+
+    for(size_t M = 0; M < nshell; M++) {
+        int nm = primary_->shell(M).nfunction();
+        int mstart = primary_->shell(M).function_index();
+        for(size_t N = 0; N < nshell; N++) {
+            int nn = primary_->shell(N).nfunction();
+            int nstart = primary_->shell(N).function_index();
+            for(size_t jki = 0; jki < njk; jki++) {
+                auto Dp = D[jki]->pointer();
+                for(size_t m = mstart; m < mstart + nm; m++) {
+                    for(size_t n = nstart; n < nstart + nn; n++) {
+                        Dshellp[M][N] = std::max(Dshellp[M][N], std::abs(Dp[m][n]));
+                    }
+                }
+            }
+        }
+    }
+
+    timer_off("Setup");
+
+    //  => First Contraction <= //
+
+    // contract D with three-index DF ERIs to get G:
+    // G_{p} = D_{mn} * (mn|p)
+
+    timer_on("ERI1");
+#pragma omp parallel for schedule(guided) num_threads(nthreads_) reduction(+ : computed_triplets1)
+    for (size_t MNP = 0; MNP < nshelltriplet; MNP++) {
+
+        size_t MN = MNP % nshellpair;
+        size_t P = MNP / nshellpair;
+        auto bra = eri_computers_[0]->shell_pairs()[MN];
+        size_t M = bra.first;
+        size_t N = bra.second;
+        if(Dshellp[M][N] * Dshellp[M][N] * J_metric_shell_diag[P] * eri_computers_[0]->shell_pair_value(M,N) < thresh2) {
```

Comment:
I would move the rank declaration to the top and use `rank` as index here as well.